### PR TITLE
Feature/chat skillstat rolls

### LINKED ISSUE #439: Improve skill/SPECIAL check rolls
- [ ] Add blind rolls
- [ ] Improve chat cards
- [ ] Add aimed shots (+2 AP, +10 on roll)
- [ ] Make attack rolls use tokens instead
- [ ] #72

### PR DESCRIPTION
**Pull request by sapphie132**
Closes #439

### CODE REVIEW

File: src/main/sass/wasteland-ventures/chat-message/system-message/weapon-attack.sass [hunk 1]
```diff
@@ -1,6 +1,6 @@
 @use "../../../constants.sass" as constants
 
-.wasteland-ventures.chat-message.system-message.weapon-attack
+.wasteland-ventures.chat-message.system-message.weapon-attack, .wasteland-ventures.chat-message.system-message.general-roll
```
Comment (kmoschcau) [inline src/main/sass/wasteland-ventures/chat-message/system-message/weapon-attack.sass:3]: I think we shouldn't just lump this together in a file called "weapon-attack.sass". Common stuff should be put in a common class and that then put on both chat cards.

File: src/main/typescript/hooks/renderChatMessage/decorateSystemMessage/index.ts [hunk 2]
```diff
@@ -23,6 +24,8 @@ export default function decorateSystemMessage(
     case "painThreshold":
       decoratePTMessage(systemFlags as PainThresholdFlags, html);
       break;
+    case "roll":
+      decorateCheck(systemFlags as CheckFlags, html);
```
Comment (kmoschcau) [inline src/main/typescript/hooks/renderChatMessage/decorateSystemMessage/index.ts:28]: Can you check if those `as` casts are still necessary or if we have some more safe way to narrow this down, please?

File: src/main/typescript/actor/wvActor.ts [hunk 3]
```diff
@@ -420,6 +426,40 @@ export default class WvActor extends Actor {
...
+      success: criticals.success.total,
+      failure: criticals.failure.total
+    });
+    const checkRoll = new Roll(fullFormula.toString()).roll({ async: false });
+    await diceSoNice(checkRoll, commonData.whisper ?? null, { actor: this.id });
+    const result = checkRoll.dice[0]?.results[0]?.result ?? 0;
+    const flags: CheckFlags = {
+      type: "roll",
+      details: {
+        criticals, // Is this an automatic type cast? Or is it a result of structural typing?
```
Comment (kmoschcau) [inline src/main/typescript/actor/wvActor.ts:446]: Yeah I think this is because of structural typing. You should probably call `toObject()` on all of the Foundry serializable composite numbers.

**Review by kmoschcau — CHANGES_REQUESTED** [review]

File: src/main/handlebars/prompt.hbs [hunk 4]
```diff
@@ -24,5 +24,23 @@
       </label>
     {{~/if~}}
   {{~/each~}}
```
Comment (kmoschcau) [inline src/main/handlebars/prompt.hbs:26]: The stuff above should probably be removed then.

File: src/main/lang/de.json [hunk 5]
```diff
@@ -456,6 +459,10 @@
         "name": "Name",
         "newName": "{what} (Neu)",
         "notes": "Notizen",
+        "rollMode": {
+          "plural": "Rollmoden",
```
Comment (kmoschcau) [inline src/main/lang/de.json:463]:
```suggestion
          "plural": "Rollmodi",
```

File: src/main/handlebars/chatMessages/check.hbs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+      </tfoot>
+    </table>
+  </div>
+</div>
+<div class="lighter-block detail-section">
+  <div>
+    <div class="roll-summary">
+      {{#if template.blinded}}
+      <label class="roll-summary-label">???</label>
+      <div class="roll-summary-result">???</div>
```
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/check.hbs:43]:
```suggestion
        <label class="roll-summary-label">???</label>
        <div class="roll-summary-result">???</div>
```

File: src/main/handlebars/chatMessages/check.hbs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      {{else}}
+      <label class="roll-summary-label">{{localize template.keys.success}}</label>
+      <div class="roll-summary-result">{{roll.result}}</div>
```
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/check.hbs:46]:
```suggestion
        <label class="roll-summary-label">{{localize template.keys.success}}</label>
        <div class="roll-summary-result">{{roll.result}}</div>
```

File: src/main/handlebars/chatMessages/check.hbs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      {{/if}}
+    </div>
+    {{#unless template.blinded}}
+    <div class="roll-summary">
+      <label class="roll-summary-label">{{localize template.keys.degrees}}</label>
+      <div class="roll-summary-result">{{roll.degree}}</div>
+    </div>
```
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/check.hbs:53]:
```suggestion
      <div class="roll-summary">
        <label class="roll-summary-label">{{localize template.keys.degrees}}</label>
        <div class="roll-summary-result">{{roll.degree}}</div>
      </div>
```

File: src/main/handlebars/chatMessages/check.hbs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    {{/unless}}
+  </div>
+  <div class="detail-content hide">
+    <div class="roll-details">
+      <div class="dice-formula">{{roll.formula}}</div>
+      <ul class="dice-rolls">
+        {{#if template.blinded}}
+        <li class="roll d100">???</li>
```
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/check.hbs:61]:
```suggestion
          <li class="roll d100">???</li>
```

File: src/main/handlebars/chatMessages/check.hbs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        {{else}}
+        <li class="roll d100">{{roll.result}}</li>
```
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/check.hbs:63]:
```suggestion
          <li class="roll d100">{{roll.result}}</li>
```

File: src/main/handlebars/chatMessages/weaponAttack.hbs [hunk 11]
```diff
@@ -118,8 +118,13 @@
   {{!hit roll}}
   <div class="lighter-block detail-section">
     <div class="roll-summary">
+      {{#if template.blinded}}
+      <label class="roll-summary-label">???</label>
+      <div class="roll-summary-result">???</div>
```
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/weaponAttack.hbs:123]:
```suggestion
        <label class="roll-summary-label">???</label>
        <div class="roll-summary-result">???</div>
```

File: src/main/handlebars/chatMessages/weaponAttack.hbs [hunk 12]
```diff
@@ -118,8 +118,13 @@
...
+      {{else}}
       <label class="roll-summary-label">{{localize template.keys.hit}}</label>
       <div class="roll-summary-result">{{rolls.hit.result}}</div>
+      {{/if}}
```
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/weaponAttack.hbs:127]:
```suggestion
      {{else}}
        <label class="roll-summary-label">{{localize template.keys.hit}}</label>
        <div class="roll-summary-result">{{rolls.hit.result}}</div>
      {{/if}}
```

File: src/main/handlebars/chatMessages/weaponAttack.hbs [hunk 13]
```diff
@@ -135,14 +140,22 @@
   <div class="lighter-block detail-section">
     <div class="roll-summary">
       <label class="roll-summary-label">{{localize "wv.rules.damage.long"}}</label>
+      {{#if template.blinded}}
+      <div class="roll-summary-result">???</div>
+      {{else}}
       <div class="roll-summary-result">{{rolls.damage.total}}</div>
+      {{/if}}
```
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/weaponAttack.hbs:147]:
```suggestion
      {{#if template.blinded}}
        <div class="roll-summary-result">???</div>
      {{else}}
        <div class="roll-summary-result">{{rolls.damage.total}}</div>
      {{/if}}
```

File: src/main/handlebars/chatMessages/weaponAttack.hbs [hunk 14]
```diff
@@ -135,14 +140,22 @@
...
     <div class="detail-content hide">
       <div class="roll-details">
         <div class="dice-formula">{{rolls.damage.formula}}</div>
         <ul class="dice-rolls">
           {{#each template.damage.results as |result|}}
+          {{#if ../template.blinded}}
+            <li class="roll d6">???</li>
+          {{else}}
             <li class="roll d6 {{result.class}}">{{result.value}}</li>
+          {{/if}}
```
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/weaponAttack.hbs:158]:
```suggestion
            {{#if ../template.blinded}}
              <li class="roll d6">???</li>
            {{else}}
              <li class="roll d6 {{result.class}}">{{result.value}}</li>
            {{/if}}
```

File: src/main/handlebars/prompt.hbs [hunk 15]
```diff
@@ -1,28 +1,35 @@
...
+    <input name="aim" type="checkbox" {{checked defaults.aim}}/>
+  </label>
+  <label class="contained-target-inline">
+    {{localize "wv.rules.actions.attack.calledShot"}}
+    <input name="calledShot" type="checkbox" {{checked defaults.calledShot}}/>
+  </label>
+  <label class="contained-target-inline">
+    {{localize "wv.rules.actions.attack.sneakAttack"}}
+    <input name="sneakAttack" type="checkbox" {{checked defaults.sneakAttack}}/>
+  </label>
```
Comment (kmoschcau) [inline src/main/handlebars/prompt.hbs:26]:
```suggestion
    <label class="contained-target-inline">
      {{localize "wv.rules.range.distance.name"}}
      <input required name="range" type="number" value="{{defaults.range}}" class="size-3" />
    </label>
    <label class="contained-target-inline">
      {{localize "wv.rules.actions.attack.aim"}}
      <input name="aim" type="checkbox" {{checked defaults.aim}}/>
    </label>
    <label class="contained-target-inline">
      {{localize "wv.rules.actions.attack.calledShot"}}
      <input name="calledShot" type="checkbox" {{checked defaults.calledShot}}/>
    </label>
    <label class="contained-target-inline">
      {{localize "wv.rules.actions.attack.sneakAttack"}}
      <input name="sneakAttack" type="checkbox" {{checked defaults.sneakAttack}}/>
    </label>
```

File: src/main/handlebars/stringPrompt.hbs [hunk 16]
```diff
@@ -0,0 +1,7 @@
+<form>
+  <label class="contained-target-inline">
+    {{label}}
+    <input required name="value" type="string" value="{{default}}" />
+  </label>
+  <button>{{localize "wv.system.actions.submit"}}</button>
+</form>
```
Comment (kmoschcau) [inline src/main/handlebars/stringPrompt.hbs:7]: The final newline is missing.

File: src/main/lang/de.json [hunk 17]
```diff
@@ -6,6 +6,13 @@
         "short": "AP",
         "use": "AP-Kosten"
       },
+      "actions": {
+        "attack": {
+          "aim": "Zielen",
+          "calledShot": "Angekündigter Schuss",
```
Comment (kmoschcau) [inline src/main/lang/de.json:12]: This doesn't seem like a great translation. I keep coming back to "Gezielter Schuss", but that could be confused with aiming in general.

File: src/main/typescript/actor/wvActor.ts [hunk 18]
```diff
@@ -12,11 +12,17 @@ import {
...
-import type { CompositeResource } from "../data/common.js";
+import type {
+  CompositeResource,
+  SerializedCompositeNumber
+} from "../data/common.js";
 import { RaceDataSourceData } from "../data/item/race/source.js";
 import Formulator, { RollOptions } from "../formulator.js";
 import { getGame } from "../foundryHelpers.js";
+import { createDefaultMessageData } from "../helpers.js";
+import type { CheckFlags as CheckFlags } from "../hooks/renderChatMessage/decorateSystemMessage/decorateCheck.js";
```
Comment (kmoschcau) [inline src/main/typescript/actor/wvActor.ts:23]: This alias is superfluous.

File: src/main/typescript/actor/wvActor.ts [hunk 19]
```diff
@@ -476,58 +482,82 @@ export default class WvActor extends Actor {
     });
   }
 
+  async rollCreateCheckMessage(
```
Comment (kmoschcau) [inline src/main/typescript/actor/wvActor.ts:485]: This is missing documentation.

Comment (kmoschcau) [reply src/main/typescript/actor/wvActor.ts:485]: Also it could probably be renamed to be more clear. For example: "createCheckRollMessage"

File: src/main/typescript/actor/wvActor.ts [hunk 20]
```diff
@@ -476,58 +482,82 @@ export default class WvActor extends Actor {
...
+    const checkRoll = new Roll(fullFormula.toString()).roll({ async: false });
+
+    const msgOptions = createDefaultMessageData(
+      ChatMessage.getSpeaker({ actor: this }),
+      options?.rollMode ?? getGame().settings.get("core", "rollMode")
+    );
+
+    msgOptions.flavor = flavor;
+
+    await diceSoNice(checkRoll, msgOptions.whisper, { actor: this.id });
```
Comment (kmoschcau) [inline src/main/typescript/actor/wvActor.ts:505]: This might also need a check for blind rolls. Right now I think a user would see the 3D dice roll animation if the roll was blind.

File: src/main/typescript/actor/wvActor.ts [hunk 21]
```diff
@@ -476,58 +482,82 @@ export default class WvActor extends Actor {
...
    * @param name - the name of the SPECIAL to roll
    */
   rollSpecial(name: SpecialName, options?: RollOptions): void {
-    const msgOptions: ConstructorDataType<foundry.data.ChatMessageData> = {
-      flavor: WvI18n.getSpecialRollFlavor(name),
-      speaker: ChatMessage.getSpeaker({ actor: this })
+    const targetRaw = this.data.data.specials[name];
+    const targetCompNum: SerializedCompositeNumber = {
+      source: targetRaw.permTotal,
+      components: targetRaw.tempComponents
```
Comment (kmoschcau) [inline src/main/typescript/actor/wvActor.ts:539]: Wouldn't it be better to use the `source` and `[...targetRaw.permComponents, ...targetRaw.tempComponents]` here?

File: src/main/typescript/applications/prompt.ts [hunk 22]
```diff
@@ -1,143 +1,48 @@
...
-      ).render(true);
-    });
-  }
-
+export abstract class RollPrompt extends Application {
   static override get defaultOptions(): ApplicationOptions {
     const defaultOptions = super.defaultOptions;
     defaultOptions.classes.push(...[CONSTANTS.systemId, "prompt"]);
     defaultOptions.template = `${CONSTANTS.systemPath}/handlebars/prompt.hbs`;
+
```
Comment (kmoschcau) [inline src/main/typescript/applications/prompt.ts:14]: Why was this added? It doesn't hurt, but I also don't see any benefit. :D

File: src/main/typescript/applications/prompt.ts [hunk 23]
```diff
@@ -1,143 +1,48 @@
...
    * @param onClose - the callback to be executed once the prompt is closed
    *                  without being submitted
    * @param options - the options for the prompt
    */
   constructor(
-    specs: Specs,
-    onSubmit: Callback<Specs>,
+    onSubmit: (data: never) => void,
     onClose: () => void,
+    data: RollPromptConstructorData,
```
Comment (kmoschcau) [inline src/main/typescript/applications/prompt.ts:31]: This parameter is undocumented.

File: src/main/typescript/applications/prompt.ts [hunk 24]
```diff
@@ -177,143 +78,237 @@ export default class Prompt<Specs extends InputSpecs> extends Application {
...
+  }
 
-/** A common input specification for a Prompt */
-export interface CommonInputSpec {
-  /** The label for the input */
-  label: string;
+  protected abstract _onSubmitCallback(data: PromptDataCommon): void;
+}
+
+export class CheckPrompt extends RollPrompt {
```
Comment (kmoschcau) [inline src/main/typescript/applications/prompt.ts:136]: This should be documented.

File: src/main/typescript/applications/prompt.ts [hunk 25]
```diff
@@ -177,143 +78,237 @@ export default class Prompt<Specs extends InputSpecs> extends Application {
...
+
+  protected override _onSubmitCallback(data: ExternalCheckData): void {
+    this.onSubmitCallback(data);
+  }
 }
 
-/** A number input specification for a Prompt  */
-export interface NumberInputSpec extends CommonInputSpec {
-  type: "number";
+export class AttackPrompt extends RollPrompt {
```
Comment (kmoschcau) [inline src/main/typescript/applications/prompt.ts:171]: This should be documented.

File: src/main/typescript/applications/prompt.ts [hunk 26]
```diff
@@ -177,143 +78,237 @@ export default class Prompt<Specs extends InputSpecs> extends Application {
...
+  protected override _onSubmitCallback(data: AttackPromptData): void {
+    this.onSubmitCallback(data);
+  }
 }
 
-/** A type that maps InputSpecs to their corresponding return types */
-export type InputSpecsReturnType<Specs extends InputSpecs> = {
-  [Key in keyof Specs]: InputSpecReturnType<Specs[Key]>;
-};
+export class StringPrompt extends Application {
```
Comment (kmoschcau) [inline src/main/typescript/applications/prompt.ts:231]: This should be documented.

File: src/main/typescript/applications/prompt.ts [hunk 27]
```diff
@@ -177,143 +78,237 @@ export default class Prompt<Specs extends InputSpecs> extends Application {
...
+        (data: string) => resolve(data),
+        reject,
+        label,
+        defaultValue ?? "",
+        options
+      ).render(true);
+    });
+  }
 
+  constructor(
```
Comment (kmoschcau) [inline src/main/typescript/applications/prompt.ts:264]: This should be documented.

File: src/main/typescript/constants.ts [hunk 28]
```diff
@@ -99,6 +99,12 @@ export function getMagicType(school: GeneralMagicSchool): MagicType {
   throw new Error("Unreachable Code");
 }
 
+export type RollMode = ValueOf<typeof CONST.DICE_ROLL_MODES>;
+
+export function isRollMode(arg: string): arg is RollMode {
+  return Object.values<string>(CONST.DICE_ROLL_MODES).includes(arg);
+}
```
Comment (kmoschcau) [inline src/main/typescript/constants.ts:106]: These should be documented.

File: src/main/typescript/data/actor/character/properties.ts [hunk 29]
```diff
@@ -36,6 +36,22 @@ export class CriticalsProperties {
...
+  constructor(apCost: number, rollMod: number) {
+    this.apCost = new CompositeNumber(apCost);
+    this.rollMod = new CompositeNumber(rollMod);
+  }
+
+  /** How much this action modifies the attack's AP cost */
+  apCost: CompositeNumber;
+
+  /** How much this action modifies the attack's roll. This can either modify
+   * crit chance or hit chance.  */
```
Comment (kmoschcau) [inline src/main/typescript/data/actor/character/properties.ts:49]:
```suggestion
  /**
   * How much this action modifies the attack's roll. This can either modify
   * crit chance or hit chance.
   */
```

File: src/main/typescript/helpers.ts [hunk 30]
```diff
@@ -14,3 +18,35 @@ export function toFixed(
...
+  const self = getGame().user;
+  switch (rollMode) {
+    case "publicroll":
+    case "blindroll":
+      return null;
+    case "gmroll":
+      return ChatMessage.getWhisperRecipients("gm");
+    case "selfroll":
+      return self ? [self] : [];
+  }
```
Comment (kmoschcau) [inline src/main/typescript/helpers.ts:51]: I am not sure if this is correct. This looks like you can't whisper a public roll to a subset of users.

File: src/main/handlebars/chatMessages/check.hbs [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+<div class="lighter-block detail-section">
+  <div>
+    <div class="roll-summary">
+      {{#if template.blinded}}
+      <label class="roll-summary-label">???</label>
+      <div class="roll-summary-result">???</div>
+      {{else}}
+      <label class="roll-summary-label">{{localize template.keys.success}}</label>
+      <div class="roll-summary-result">{{roll.result}}</div>
+      {{/if}}
```
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/check.hbs:47]: To DRY up code, you could also use the `ternary` helper instead.

File: src/main/typescript/hooks/renderChatMessage/decorateSystemMessage/decorateCheck.ts [hunk 32]
```diff
@@ -0,0 +1,80 @@
...
+  return;
+}
+
+export type CheckFlags = CommonRollFlags & {
+  type: "roll";
+  roll: {
+    formula: string;
+    critical?: Critical | undefined;
+    result: number;
+    degree: number;
```
Comment (kmoschcau) [inline src/main/typescript/hooks/renderChatMessage/decorateSystemMessage/decorateCheck.ts:41]: This should probably be called "degrees" or even "degreesOfSuccess".

File: src/main/typescript/integrations/quench/registerBatches.ts [hunk 33]
```diff
@@ -1,8 +1,8 @@
-import { CONSTANTS } from "../../constants.js";
-import type { HookParams } from "../../hooks/quench/ready.js";
-import registerPromptTests from "../../tests/applications/prompt.js";
+// import { CONSTANTS } from "../../constants.js";
```
Comment (kmoschcau) [inline src/main/typescript/integrations/quench/registerBatches.ts:1]: This file should be deleted instead of commented out.

File: src/main/typescript/wvI18n.ts [hunk 34]
```diff
@@ -56,6 +57,9 @@ export type I18nRadSicknessLevels = Record<RadiationSicknessLevel, string>;
 /** The internationalization structure for rarities */
 export type I18nRarities = Record<Rarity, string>;
 
+/** The internationalisation structure for rollmodes */
```
Comment (kmoschcau) [inline src/main/typescript/wvI18n.ts:60]:
```suggestion
/** The internationalization structure for roll modes */
```

File: src/main/typescript/wvI18n.ts [hunk 35]
```diff
@@ -163,6 +167,19 @@ export default class WvI18n {
...
+  /** Get the internationalization of the Dice Roll Modes */
+  static get rollModes(): I18nRollModes {
+    const base = "CHAT.Roll";
+    return Object.entries(CONST.DICE_ROLL_MODES).reduce((acc, [key, value]) => {
+      const internalisationKey = base + key.toLowerCase().capitalize();
+      acc[value] = foundry.utils.getProperty(
+        getGame().i18n.translations,
+        internalisationKey
+      );
+      return acc;
```
Comment (kmoschcau) [inline src/main/typescript/wvI18n.ts:179]:
```suggestion
      const internalizationKey = base + key.toLowerCase().capitalize();
      acc[value] = foundry.utils.getProperty(
        getGame().i18n.translations,
        internalizationKey
      );
      return acc;
```

**Review by kmoschcau — CHANGES_REQUESTED** [review]

File: src/main/typescript/integrations/quench/registerBatches.ts (hunk 33 above)
Comment (sapphie132) [reply src/main/typescript/integrations/quench/registerBatches.ts:1]: Oh, I knew I was forgetting something. Done.

File: src/main/handlebars/chatMessages/check.hbs [hunk 36]
```diff
@@ -0,0 +1,59 @@
...
+    <div class="roll-summary">
+      <label class="roll-summary-label">{{ternary template.blinded "???" (localize template.keys.success)}}</label>
+      <div class="roll-summary-result">{{ternary template.blinded "???" roll.result}}</div>
+    </div>
+    {{#unless template.blinded}}
+    <div class="roll-summary">
+      <label class="roll-summary-label">{{localize template.keys.degrees}}</label>
+      <div class="roll-summary-result">{{roll.degree}}</div>
+    </div>
+    {{/unless}}
```
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/check.hbs:49]:
```suggestion
    {{#unless template.blinded}}
      <div class="roll-summary">
        <label class="roll-summary-label">{{localize template.keys.degrees}}</label>
        <div class="roll-summary-result">{{roll.degree}}</div>
      </div>
    {{/unless}}
```

File: src/main/handlebars/chatMessages/weaponAttack.hbs (hunk 12 above)
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/weaponAttack.hbs:127]:
```suggestion
      <label class="roll-summary-label">{{ternary template.blinded "???" (localize template.keys.hit)}}</label>
      <div class="roll-summary-result">{{ternary template.blinded "???" rolls.hit.result}}</div>
```

File: src/main/handlebars/chatMessages/weaponAttack.hbs (hunk 13 above)
Comment (kmoschcau) [inline src/main/handlebars/chatMessages/weaponAttack.hbs:147]:
```suggestion
      <div class="roll-summary-result">{{ternary template.blinded "???" rolls.damage.total}}</div>
```

File: src/main/handlebars/prompt.hbs [hunk 37]
```diff
@@ -1,28 +1,35 @@
...
+  {{/if}}
```
Comment (kmoschcau) [inline src/main/handlebars/prompt.hbs:27]:
```suggestion
  {{#if isAttack}}
    <label class="contained-target-inline">
      {{localize "wv.rules.range.distance.name"}}
      <input required name="range" type="number" value="{{defaults.range}}" class="size-3" />
    </label>
    <label class="contained-target-inline">
      {{localize "wv.rules.actions.attack.aim"}}
      <input name="aim" type="checkbox" {{checked defaults.aim}}/>
    </label>
    <label class="contained-target-inline">
      {{localize "wv.rules.actions.attack.calledShot"}}
      <input name="calledShot" type="checkbox" {{checked defaults.calledShot}}/>
    </label>
    <label class="contained-target-inline">
      {{localize "wv.rules.actions.attack.sneakAttack"}}
      <input name="sneakAttack" type="checkbox" {{checked defaults.sneakAttack}}/>
    </label>
  {{/if}}
```

File: src/main/typescript/applications/item/weaponSheet.ts [hunk 38]
```diff
@@ -24,6 +24,7 @@ export default class WeaponSheet extends WvItemSheet {
     defaultOptions.height = 500;
     defaultOptions.width = 800;
     return foundry.utils.mergeObject(defaultOptions, {
+      // TODO? Should this be changed too? Having a tough time understanding what this code does
```
Comment (kmoschcau) [inline src/main/typescript/applications/item/weaponSheet.ts:27]: This determines which HTML elements can be dragged on the sheet. This was used in the past to allow dragging weapon attack buttons into the hotbar. Though we would need to refactor this anyway, because for players at least we would need to add a check that the weapon is also readied when they execute the attack.

File: src/main/typescript/hooks/index.ts [hunk 39]
```diff
@@ -2,7 +2,7 @@ import registerForDragRulerReady from "./dragRuler/ready.js";
 import registerForHotbarDrop from "./hotbarDrop.js";
 import registerForInit from "./init.js";
 import registerForPreUpdateToken from "./preUpdateToken.js";
-import registerForQuenchReady from "./quench/ready.js";
+// import registerForQuenchReady from "./quench/ready.js";
```
Comment (kmoschcau) [inline src/main/typescript/hooks/index.ts:5]: This should be removed.

File: src/main/typescript/hooks/index.ts [hunk 40]
```diff
@@ -22,5 +22,5 @@ export default function registerForHooks(): void {
   registerForUpdateActor();
 
   registerForDragRulerReady();
-  registerForQuenchReady();
+  // registerForQuenchReady();
```
Comment (kmoschcau) [inline src/main/typescript/hooks/index.ts:25]: This should be removed.

File: src/main/typescript/wvI18n.ts [hunk 41]
```diff
@@ -163,6 +167,19 @@ export default class WvI18n {
     ) as I18nSpecials;
   }
 
+  /** Get the internationalization of the Dice Roll Modes */
```
Comment (kmoschcau) [inline src/main/typescript/wvI18n.ts:170]:
```suggestion
  /** Get the internationalization of the Dice Roll Modes. */
```

**Review by kmoschcau — CHANGES_REQUESTED** [review]

File: src/main/typescript/applications/item/weaponSheet.ts (hunk 38 above)
Comment (sapphie132) [reply src/main/typescript/applications/item/weaponSheet.ts:27]: But for now I should still change the html class, correct?

File: src/main/lang/de.json (hunk 17 above)
Comment (sapphie132) [reply src/main/lang/de.json:12]: That's exactly why I picked this one instead. It's clunky, but it seems like the only option to keep things unambiguous

File: src/main/typescript/actor/wvActor.ts (hunk 21 above)
Comment (sapphie132) [reply src/main/typescript/actor/wvActor.ts:557]: Huh, I could have sworn `permComponents` was included in `tempComponents`.

File: src/main/typescript/applications/prompt.ts (hunk 22 above)
Comment (sapphie132) [reply src/main/typescript/applications/prompt.ts:14]: Don't remember. _Might_ have been the linter/prettier at some point, but I doubt it. Should I remove it?

File: src/main/typescript/helpers.ts (hunk 30 above)
Comment (sapphie132) [reply src/main/typescript/helpers.ts:51]: Is that a base feature in foundry? Because, as far as I know, a public roll is supposed to be completely public. I've never thought of the possibility of having a roll be private to a certain group of people.

Comment (kmoschcau) [reply src/main/typescript/helpers.ts:51]: Well those are two separate privacy settings. A public message is seen by everyone. A whispered message is only seen by the creator and the message recipients.

A public roll shows the result to everyone that can see the message. A blind roll shows the result only to GMs. So it is a valid combination.

File: src/main/typescript/applications/item/weaponSheet.ts (hunk 38 above)
Comment (kmoschcau) [reply src/main/typescript/applications/item/weaponSheet.ts:27]: You still need to change the CSS class correct. But this should not be affected by this change. The drag selectors are from the point of view of the application element root, not from the overall page document root.

This means that the selector here is already inside of the element that is affected by your CSS changes.

File: src/main/typescript/actor/wvActor.ts [hunk 42]
```diff
@@ -476,58 +485,97 @@ export default class WvActor extends Actor {
     });
   }
 
+  /** Roll a skill or SPECIAL check and post the result
```
Comment (kmoschcau) [inline src/main/typescript/actor/wvActor.ts:488]:
```suggestion
  /**
   * Roll a skill or SPECIAL check and post the result
```

**Review by kmoschcau — APPROVED** [review]
